src/service/resources-impl.js

```diff
@@ -281,94 +287,144 @@ export class ResourcesImpl {
 
   /** @override */
   isIntersectionExperimentOn() {
-    return !!this.intersectionObserver_;
+    return !!this.oneViewportObserver_;
   }
 
   /**
    * @param {!Array<!IntersectionObserverEntry>} entries
-   * @param {!IntersectionObserver} unusedObserver
    * @private
    */
-  intersects_(entries, unusedObserver) {
-    dev().fine(TAG_, 'intersect', entries);
+  intersectsOneViewport_(entries) {
+    // TODO(willchou): Remove assert once #27167 is fixed.
+    devAssert(this.prerenderSize_ == 1);
+
+    // Store the 1vp rect for later.
+    if (!this.oneViewportRect_) {
+      this.oneViewportRect_ =
+        /** @type {!ClientRect} */ (entries[0].rootBounds);
+    }
 
-    const toUnload = [];
+    const whenBuilt = this.intersects_(entries, /* viewports */ 1);
+
+    // To prioritize in-viewport elements, delay handling of 2vp/3vp until
+    // all 1vp elements are built (and we're not in prerender, of course).
+    Promise.all([whenBuilt, this.ampdoc.whenFirstVisible()]).then(() => {
+      if (this.threeViewportsObserver_) {
+        return;
+      }
+      this.threeViewportsObserver_ = new IntersectionObserver(
+        e => this.intersectsThreeViewports_(e),
+        {
+          root: this.oneViewportObserver_.root,
+          rootMargin: '200% 25%',
+        }
+      );
+      const elements = this.resources_.map(r => r.element);
+      elements.forEach(el => {
+        el.whenUpgraded().then(() => this.threeViewportsObserver_.observe(el));
+      });
+    });
+  }
+
+  /**
+   * @param {!Array<!IntersectionObserverEntry>} entries
+   * @private
+   */
+  intersectsThreeViewports_(entries) {
+    devAssert(this.visible_);
+
+    // Always ignore intersections in 1vp to avoid dupe handling.
+    const outsideOneViewport = entries.filter(
+      e =>
+        !rectsOverlap(
+          devAssert(this.oneViewportRect_),
+          /** @type {!ClientRect} */ (e.intersectionRect)
+        )
+    );
 
+    const whenBuilt = this.intersects_(outsideOneViewport, /* viewports */ 3);
+
+    // Defer building remaining elements on page (4vp+) until 3vp completes.
+    // Note: these elements no longer receive onLayoutMeasure/onMeasureChanged
+    // until an intersection is observed (3vp or closer).
+    whenBuilt.then(() => {
+      for (let i = 0; i < this.resources_.length; i++) {
+        const r = this.resources_[i];
+        if (r.getState() == ResourceState.NOT_BUILT && !r.isBuilding()) {
+          this.buildOrScheduleBuildForResource_(r, /* checkForDupes */ true);
```

Comment:
We are blocking build? We've not blocked build before - only loading. This is certain to be positive, but might have lots of regressions with the elements that do some early work on build?

Also, does the `whenBuilt` never fail for any individual elements, or all together? Overall this seems like a pretty strict rule. What if a 1vp element is very slow in download?

Reply:
Yea, I was originally planning to split this out but I think this should be safe to do since we already do something similar in prerender mode with our quota system. 

Good points on being bottlenecked by slow outliers. I'll think about how to mitigate.